src/main/scala/dsptools/numbers/algebra_types/Signed.scala

```diff
@@ -42,6 +43,11 @@ object Signed {
 
 private class OrderedRingIsSigned[A <: Data](implicit o: Order[A], r: Ring[A]) extends Signed[A] {
   def signum(a: A): ComparisonBundle = o.compare(a, r.zero)
-  def abs(a: A): A = Mux(signum(a).lt, r.negate(a), a)
+  def abs(a: A): A = {
+    Mux(signum(a).lt, r.negate(a), a)  //TODO: should this be negateContext
+  }
+  def context_abs(a: A): A = {
```

Comment:
Check out what I did with the 3 multiply version of complex multiply. I had to do something like (I forget the ops) a + c*d but since c *d is delayed by numMulPipes, I needed to delay a by the same amount. Basically you should use shiftregister(a, context.numAddPipes) instead of a. And likewise for the condition. You should be able to catch delay misalignments if you test contexts with delays. That's actually probably pretty important.